RealmSwift/RealmKeyedCollection.swift

```diff
@@ -308,6 +308,87 @@ public protocol RealmKeyedCollection: Sequence, ThreadConfined, CustomStringConv
+ enables aggregatable operations.
+ */
+public protocol KeyedAggregatable: RealmKeyedCollection where Value: OptionalProtocol, Value.Wrapped: ObjectBase, Value.Wrapped: RealmCollectionValue {
+    /**
+     Returns the minimum (lowest) value of the given property among all the objects in the collection, or `nil` if the
+     collection is empty.
+
+     - warning: Only a property whose type conforms to the `MinMaxType` protocol can be specified.
+
+     - parameter property: The name of a property whose minimum value is desired.
+     */
+    func min<T: MinMaxType>(ofProperty property: KeyPath<Value.Wrapped, T>) -> T?
+
+    /**
+     Returns the maximum (highest) value of the given property among all the objects in the collection, or `nil` if the
+     collection is empty.
+
+     - warning: Only a property whose type conforms to the `MinMaxType` protocol can be specified.
+
+     - parameter property: The name of a property whose minimum value is desired.
+     */
+    func max<T: MinMaxType>(ofProperty property: KeyPath<Value.Wrapped, T>) -> T?
+
+    /**
+    Returns the sum of the given property for objects in the collection, or `nil` if the collection is empty.
+
+    - warning: Only names of properties of a type conforming to the `AddableType` protocol can be used.
+
+    - parameter property: The name of a property conforming to `AddableType` to calculate sum on.
+    */
+    func sum<T: AddableType>(ofProperty property: KeyPath<Value.Wrapped, T>) -> T
+
+    /**
+     Returns the average value of a given property over all the objects in the collection, or `nil` if
+     the collection is empty.
+
+     - warning: Only a property whose type conforms to the `AddableType` protocol can be specified.
+
+     - parameter property: The name of a property whose values should be summed.
+     */
+    func average<T: AddableType>(ofProperty property: KeyPath<Value.Wrapped, T>) -> T?
+}
+
+// MARK: Sortable
+
+/**
+ Protocol for RealmKeyedCollections where the Value is of an Object type that
+ enables sortable operations.
+ */
+public protocol KeyedSortable: RealmKeyedCollection where Value: OptionalProtocol, Value.Wrapped: ObjectBase, Value.Wrapped: RealmCollectionValue {
+    /**
+     Returns a `Results` containing the objects in the collection, but sorted.
+
+     Objects are sorted based on the values of the given key path. For example, to sort a collection of `Student`s from
+     youngest to oldest based on their `age` property, you might call
+     `students.sorted(byKeyPath: "age", ascending: true)`.
+
+     - warning: Collections may only be sorted by properties of boolean, `Date`, `NSDate`, single and double-precision
+                floating point, integer, and string types.
+
+     - parameter keyPath:   The key path to sort by.
+     - parameter ascending: The direction to sort in.
+     */
+    func sorted<T: Comparable>(byKeyPath keyPath: KeyPath<Value.Wrapped, T>, ascending: Bool) -> Results<Value.Wrapped>
```

Comment:
Does this drop null values when sorting? That seems like a weird thing for it to do, but unless it is doing that then the return type needs to be `Results<Value>`

Reply:
Why does that seem like a weird thing for it to do? I considered this a nicer alternative. What does a sorted Results with nil values mean semantically?

Reply:
Normally a sorted collection is always the same length as the unsorted one because sorting just rearranges values without removing any. We've taken the general position that null is less than all other values.

I assume that this will currently just crash at runtime if there's any null values since there's nothing actually filtering them out.

Reply:
I have since removed it and have gone back to returning the nullable element type.